# Practice Case Week 4 NoSQL - MongoDB
## Muhammad Ridho Aulia

Please complete the following task:
1. Create free tier MongoDB cluster to get free trial MongoDB cluster
2. Load `sample_mflix` dataset on your cluster
3. Upload `movies_initial` dataset to your cluster
4. Replicate `sample_mflix` dataset from `movies_initial` using projecting queries named as `clean_movies`. You can choose either with MongoDB or query by MongoDB Compass
5. Your script must include the following queries:
    * a. Show that All documents on `clean_movies` and `sample_mflix` are same
    * b. Show that Number of documents on both `clean_movies` and `sample_mflix` are same
    * c. Show that all fields on `clean_movies` available in sample_mflix
    * d. Show that all values on both `clean_movies` and `sample_mflix` collection are same with same order
    * e. If all the parameters has not tally, describe the reason
    
In this case, we want to create a new collection using this mongo client:

MongoClient("mongodb+srv://mra:mra1994@iykramra-wh2hy.gcp.mongodb.net/test?retryWrites=true&w=majority")

that replicate sample_mflix.movies from dataset movies_initial using projecting queries named as clean_movies.

The workflow is following:

1. Importing the modules if necessary
2. Loading dataset movies_initial on MongoDB cluster
3. Connecting to client
4. Do a little Exploratory collection in Movies, and Movies Initial
5. Do aggregation using pipeline : ' 𝑝𝑟𝑜𝑗𝑒𝑐𝑡  (for changing the field's format) -->  𝑚𝑎𝑡𝑐ℎ  (for query documents) -->  𝑜𝑢𝑡  (for create new collection)
6. Validate the result

### Installing MongoDB pymongo driver and dnspython in order to connect using pymongo SRV

In [25]:
!pip install pymongo
!pip install pymongo[srv]
!pip install dnspython

### import any useful modules, using pprint for proper print for json results 

In [75]:
import pymongo                            
from pymongo import MongoClient               
from pprint import pprint                                 
import pandas as pd                          
from datetime import datetime 
import json

In [41]:
# Make a connection to mongodb cluster

client = MongoClient("mongodb+srv://mra:mra1994@iykramra-wh2hy.gcp.mongodb.net/test?retryWrites=true&w=majority")

### Create new database for this task

In [42]:
import pymongo

myclient = pymongo.MongoClient("mongodb://localhost:27017/")

mydb = myclient["movies"]

In [49]:
#Convert movies_initial.csv to json
df = pd.read_csv (r'movies_initial.csv')
df.to_json (r'movies_initial.json')

In [66]:
# Import 'movies_initial.csv' into MongoDB using pandas DataFrame
df = pd.read_json('movies_initial.json')
df = df.fillna("") # null values treated as number in MongoDB
                   # replace with empty string
                   
mydb = client.movies
mydb.movies_initial.insert_many(df.to_dict('records'))
print("Import success")

Import success


In [67]:
# Print database names in the cluster
client.list_database_names()

['movies',
 'practice_mra',
 'sample_airbnb',
 'sample_analytics',
 'sample_geospatial',
 'sample_mflix',
 'sample_supplies',
 'sample_training',
 'sample_weatherdata',
 'video',
 'admin',
 'local']

In [68]:
# Print collection names in sample_mflix database
client.sample_mflix.list_collection_names()

['movies', 'sessions', 'comments', 'users', 'theaters']

In [95]:
# Print collection names in movies database
client.movies.list_collection_names()

['clean_movies', 'movies_initial']

In [73]:
# Save the collection into variable
movies_initial = client.movies.movies_initial
movies_mflix   = client.sample_mflix.movies

# Count the documents in each collection
movies_initial_count = movies_initial.count_documents({})
movies_mflix_count   = movies_mflix.count_documents({})

# Print the count the documents in each collection
print("movies_initial : " + str(movies_initial_count))
print("movies_mflix   : " + str(movies_mflix_count))
print("Difference     : " + str(movies_initial_count-movies_mflix_count))

movies_initial : 46014
movies_mflix   : 23539
Difference     : 22475


The difference between those collection is 22475. We will see the document and try to fix this

In [76]:
# Print the document in movies_mflix using pprint
print("movies_mflix")
pprint(movies_mflix.find_one())

movies_mflix
{'_id': ObjectId('573a1390f29313caabcd4135'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': 1,
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 'released': date

In [77]:
# Check one document from movies_initial to see the structure inside it
pprint(movies_initial.find_one({'imdbID': 5}))

{'_id': ObjectId('5eafd4809af01fcc98d179d2'),
 'awards': '1 win.',
 'cast': 'Charles Kayser, John Ott',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genre': 'Short',
 'imdbID': 5,
 'imdbRating': 6.2,
 'imdbVotes': 1189.0,
 'language': '',
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'metacritic': '',
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'poster': '',
 'rating': 'UNRATED',
 'released': '1893-05-09',
 'runtime': '1 min',
 't

There are many differences in those documents. For example like awards, cast, country(ies), director(s), genre(s), imdb (id, rating, votes), tomatoes (missing), runtime and release have different format or type.


The movies_mflix used embedded document to group field from same sources, for example:

    imdb': {'id': 5, 'rating': 6.2, 'votes': 1189}

While in movies_initial we have:

    'imdbID': 5,
    'imdbRating': 6.2,
    'imdbVotes': 1189.0,

The movies_mflix used list for key with several values, for example:

    'cast': ['Charles Kayser', 'John Ott']
While in movies_initial we have:

    'cast': 'Charles Kayser, John Ott'

In [80]:
# Create doc ID, rating, and votes from imdb field.
imdb = list(movies_mflix.find({},{'_id':0,'imdb':1}))
imdb_list=[]
for title in imdb:
  imdb_list.append(list(title.values())[0])

print(len(imdb_list))
imdb_list[:5]

23539


[{'rating': 6.2, 'votes': 1189, 'id': 5},
 {'rating': 7.4, 'votes': 9847, 'id': 439},
 {'rating': 7.1, 'votes': 448, 'id': 488},
 {'rating': 6.6, 'votes': 1375, 'id': 832},
 {'rating': 7.3, 'votes': 1034, 'id': 1737}]

## Aggregation pipeline

In [81]:
pipeline = [
        {
        # Stage 1 -> Project
        '$project': {
            
            # $cond use here so that if there are no title in movies_initial, then we use $nonexistinField that will not showing title 
            'title': {"$cond":[{"$eq":["$title",""]},"$nonExistinField","$title"]},
            'year': {"$cond":[{"$eq":["$year",""]},"$nonExistinField","$year"]},
            'plot': {"$cond":[{"$eq":["$plot",""]},"$nonExistinField","$plot"]},
            
            # Directors,cast etc is a list in movies, so we use split, that automatically split string by commas in our case
            # Then, automaticaly, the output will be on list
            
            'directors': {"$cond":[{"$eq":["$director",""]},"$nonExistinField",{'$split': ["$director", ", "]}]},
            'cast': {"$cond":[{"$eq":["$cast",""]},"$nonExistinField",{'$split': ["$cast", ", "]}]},
            'writers': {"$cond":[{"$eq":["$writer",""]},"$nonExistinField",{'$split': ["$writer", ", "]}]},
            'genres': {"$cond":[{"$eq":["$genre",""]},"$nonExistinField",{'$split': ["$genre", ", "]}]},
            'countries': {"$cond":[{"$eq":["$country",""]},"$nonExistinField",{'$split': ["$country", ", "]}]},
            
            'fullplot': {"$cond":[{"$eq":["$fullplot",""]},"$nonExistinField","$fullplot"]},
            
            # In released, the format is in datetime. We already tried using datetime.strptime but it seems $released is not readable for input in strptime
            'released': {"$cond":[{"$eq":["$released",""]},"$nonExistinField",{ '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 2]},
                                      ", 0, 0)" ] }]},
            # In movies, there are no string "min" in runtime. So we remove it
            # Condition used so that if there are empty string in documents, it assign into zero
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, "$nonExistinField" ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            
            # Metacritic remove metacritic if values in movies_initial "", otherwise show the values
            'metacritic':{"$cond":[{"$eq":["$metacritic",""]},"$nonExistinField","$metacritic"]},
            
            # The values in num_mflix_comments at movies_initial are all "" so we assign it to num_mflix_comments
            'num_mflix_comments': "",
            
            
            'poster': {"$cond":[{"$eq":["$poster",""]},"$nonExistinField","$poster"]},
            'rated': {"$cond":[{"$eq":["$rating",""]},"$nonExistinField","$rating"]},
            
            # imdb values in a movies_initial are grouped into one dictionaries in fields imdb
            'imdb': {
                'id': "$imdbID",
                'rating': "$imdbRating",
                'votes': "$imdbVotes"
                },
            
            # using split and arrayElemAt, we extract strings value in $awards at movies_initial
            'awards': {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            'type': {"$cond":[{"$eq":["$type",""]},"$nonExistinField","$type"]},
            'lastupdated': {"$cond":[{"$eq":["$lastupdated",""]},"$nonExistinField","$lastupdated"]},
            
            # Tomatoes in movies, are in format dictionaries we split $lastupdated to extract hour,minutes,second
            'tomatoes':{
            'lastupdated':{ '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$lastupdated", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$lastupdated", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ {'$arrayElemAt':[{'$split': [ "$lastupdated", " " ]}, 0]} ,"-"]}, 2]},
                                      ", ",{'$arrayElemAt':[{'$split': [ {'$arrayElemAt':[{'$split': [ "$lastupdated", " " ]}, 1]} ,":"]}, 0]},
                                      ", ",{'$arrayElemAt':[{'$split': [ {'$arrayElemAt':[{'$split': [ "$lastupdated", " " ]}, 1]} ,":"]}, 1]},
                                      ", ",{"$arrayElemAt":[{"$split":[{'$arrayElemAt':[{'$split': [ {'$arrayElemAt':[{'$split': [ "$lastupdated", " " ]}, 1]} ,":"]}, 2]},"."]},0]},
                                        ")"] },
                'viewer': {'meter':'', 'numReviews':'', 'rating':''},
                'dvd':'',
                'fresh':'',
                'rotten':'',
                'production':''},
            'languages': {'$split': ["$language", ", "]}
               
    },
    },{
         "$match":{"imdb":{"$in":[dict(sorted(imdb_list[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}}  
       },
    
    
    {
        ### naming the new collection as output ### 
        '$out': "clean_movies"
    }
]
client.movies.drop_collection("clean_movies")
movies_initial.aggregate(pipeline)
clean_movies = client.movies.clean_movies
print("Aggregation Pipeline success")

Aggregation Pipeline success


### A. Show that All documents on clean_movies and sample_mflix are same

In [82]:
pprint(clean_movies.find_one())

{'_id': ObjectId('5eafd4809af01fcc98d179d2'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189.0},
 'languages': [''],
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': '',
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 '

In [83]:
pprint(movies_mflix.find_one())

{'_id': ObjectId('573a1390f29313caabcd4135'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': 1,
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 'released': datetime.datetime

In [84]:
list_mflix_id = movies_mflix.aggregate([{ "$match": {"imdb.id": {"$gt": 0}} },
                                   { "$project": { "title": "$name",
                                                 "year": "$name"
                                             } 
                                   }])
#list_clean_id = clean_movies

### B. Show that Number of documents on both `clean_movies` and `sample_mflix` are same

In [85]:
# Count the documents in each collection
clean_movies_count = clean_movies.count_documents({})
movies_mflix_count = movies_mflix.count_documents({})

# Print the count the documents in each collection
print("clean_movies : " + str(clean_movies_count))
print("movies_mflix : " + str(movies_mflix_count))
print("Difference   : " + str(clean_movies_count - movies_mflix_count))

clean_movies : 23539
movies_mflix : 23539
Difference   : 0


### C. Show that all fields on clean_movies available in sample_mflix

In [90]:
clean= set([z for i in [doc.keys() for doc in clean_movies.find({})]for z in i])
print(clean)
print('\n'+'clean_movies total field: '+str(len(clean)))

{'rated', 'languages', 'released', 'metacritic', 'writers', 'plot', 'num_mflix_comments', 'year', 'awards', 'runtime', '_id', 'type', 'countries', 'directors', 'imdb', 'tomatoes', 'cast', 'poster', 'title', 'genres', 'lastupdated', 'fullplot'}

clean_movies total field: 22


In [91]:
mflix= set([z for i in [doc.keys() for doc in movies_mflix.find({})]for z in i])
print(mflix)
print('\n'+'sample_mflix total field: '+str(len(mflix)))

{'rated', 'languages', 'released', 'metacritic', 'writers', 'plot', 'num_mflix_comments', 'year', 'awards', 'runtime', '_id', 'type', 'countries', 'directors', 'imdb', 'tomatoes', 'cast', 'poster', 'title', 'genres', 'lastupdated', 'fullplot'}

sample_mflix total field: 22


### D. Show that all values on both clean_movies and sample_mflix collection are same with same order

In [92]:
pprint(clean_movies.find_one())

{'_id': ObjectId('5eafd4809af01fcc98d179d2'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189.0},
 'languages': [''],
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': '',
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 '

In [93]:
pprint(movies_mflix.find_one())

{'_id': ObjectId('573a1390f29313caabcd4135'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': 1,
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 'released': datetime.datetime

### E. If all the parameters has not tally, describe the reason

1. There are many fields with empty string values in clean_movies

2. Collection imdb.votes can't be converted to integer from float because there are string values in some documents